pkg/sdn/plugin/controller.go

```diff
@@ -330,6 +337,106 @@ func (plugin *OsdnNode) SetupSDN(localSubnetCIDR, clusterNetworkCIDR, servicesNe
 	return true, nil
 }
 
+func (plugin *OsdnNode) SetupEgressNetworkPolicy() error {
+	policies, err := plugin.registry.GetEgressNetworkPolicies()
+	if err != nil {
+		return fmt.Errorf("Could not get EgressNetworkPolicies: %s", err)
+	}
+
+	for _, policy := range policies {
+		vnid, err := plugin.getVNID(policy.Namespace)
+		if err != nil {
+			glog.Warningf("Could not find namespace %q for EgressNetworkPolicy: %v", err)
+			continue
+		}
+		err = plugin.UpdateEgressNetworkPolicy(policy, vnid)
+		if err != nil {
+			return err
+		}
+	}
+
+	go utilwait.Forever(plugin.watchEgressNetworkPolicies, 0)
+	return nil
+}
+
+func (plugin *OsdnNode) watchEgressNetworkPolicies() {
+	eventQueue := plugin.registry.RunEventQueue(EgressNetworkPolicies)
+
+	for {
+		eventType, obj, err := eventQueue.Pop()
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("EventQueue failed for EgressNetworkPolicy: %v", err))
+			return
+		}
+		policy := obj.(*osapi.EgressNetworkPolicy)
+
+		vnid, err := plugin.getVNID(policy.Namespace)
+		if err != nil {
+			glog.Warningf("Could not find namespace %q for EgressNetworkPolicy: %v", err)
+			continue
+		}
+
+		if eventType == watch.Deleted {
+			policy.Spec.Egress = nil
+		}
+		err = plugin.UpdateEgressNetworkPolicy(*policy, vnid)
+		if err != nil {
+			utilruntime.HandleError(err)
+			return
+		}
+	}
+}
+
+func (plugin *OsdnNode) UpdateEgressNetworkPolicy(policy osapi.EgressNetworkPolicy, vnid uint32) error {
+	if policy.Name != "default" {
+		glog.Warningf("Ignoring EgressNetworkPolicy %q in namespace %q", policy.Name, policy.Namespace)
+		return nil
+	}
+	if vnid == AdminVNID {
+		if len(policy.Spec.Egress) != 0 {
+			glog.Warningf("Ignoring EgressNetworkPolicy rules for global namespace %q", policy.Namespace)
+		}
+		return nil
+	}
+
+	otx := ovs.NewTransaction(kexec.New(), BR)
+
+	if len(policy.Spec.Egress) == 0 {
+		otx.DeleteFlows("table=9, reg0=%d", vnid)
+	} else {
+		// Temporarily drop all outgoing traffic, to avoid race conditions while modifying the other rules
+		otx.AddFlow("table=9, reg0=%d, cookie=1, priority=65535, actions=drop", vnid)
+		otx.DeleteFlows("table=9, reg0=%d, cookie=0/1", vnid)
+
+		for i, rule := range policy.Spec.Egress {
+			priority := len(policy.Spec.Egress) - i
+
+			var action string
+			if rule.Type == osapi.EgressNetworkPolicyRuleAllow {
+				action = "output:2"
+			} else {
+				action = "drop"
+			}
+
+			var dst string
+			if rule.To.CIDRSelector == "0.0.0.0/32" {
+				dst = ""
+			} else {
+				dst = fmt.Sprintf(", nw_dst=%s", rule.To.CIDRSelector)
+			}
+
+			otx.AddFlow("table=9, reg0=%d, priority=%d, ip%s, actions=%s", vnid, priority, dst, action)
```

Comment:
This has some side effects.
(1) Let's say we have 2 projects A and B
(2) Apply some egress policy to project A
(3) For some reason, project B needs to have access to pods in project A. So project networks for A and B are merged.
Since project A and B are sharing the same vnid, egress policy in (2) will be applied to project B as well which was not intended. I can't think of easy way to deal with this as source IP may not match originated pod IP. May be for now, document this limitation and tackle this case in the next release?
@knobunc @dcbw what do you think?